x/programs/cmd/simulator/src/simulator.rs

```diff
@@ -0,0 +1,238 @@
+use crate::{
+    codec::{base64_decode, id_from_usize},
+    param::Param,
+    Id,
+};
+use borsh::BorshDeserialize;
+use serde::{Deserialize, Serialize};
+use std::{
+    io::{BufRead, BufReader, Write},
+    process::{Child, Command, Stdio},
+};
+use thiserror::Error;
+use wasmlanche_sdk::ExternalCallError;
+
+#[derive(Error, Debug)]
+pub enum SimulatorResponseError {
+    #[error(transparent)]
+    Serialization(#[from] borsh::io::Error),
+    #[error(transparent)]
+    ExternalCall(#[from] ExternalCallError),
+}
+
+#[derive(Error, Debug)]
+pub enum ClientError {
+    #[error("Read error: {0}")]
+    Read(#[from] std::io::Error),
+    #[error("EOF")]
+    Eof,
+    #[error("Missing handle")]
+    StdIo,
+}
+
+#[derive(Error, Debug)]
+pub enum SimulatorError {
+    #[error("Client error: {0}")]
+    Client(#[from] ClientError),
+    #[error("Serialization / Deserialization error: {0}")]
+    Serde(#[from] serde_json::Error),
+    #[error("Borsh deserialization error: {0}")]
+    BorshDeserialization(#[from] borsh::io::Error),
+    #[error("Program error: {0}")]
+    Program(String),
+}
+
+pub struct Simulator<W, R> {
+    writer: W,
+    responses: R,
+}
+
+pub fn build_simulator() -> Simulator<impl Write, impl Iterator<Item = SimulatorResponseItem>> {
+    let path = get_binary_path();
+
+    let Child { stdin, stdout, .. } = Command::new(path)
+        .arg("interpreter")
```

Comment:
Have you considered making the interpreter mode a default ? Since we don't use the subcommands natively at all it may simplify some stuff with these subcommands but if that is a good choice that could probably be better as a followup of this pr

Reply:
yea I agree, need to look into it more was out of the scope of this PR since its getting big already. noted for a follow up
https://github.com/ava-labs/hypersdk/issues/1169